File: bench/wordfreq/wordfreq.js
// Word frequency counting: string split, dict updates, and a sort with a key.
const vocab = ["alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta"];

const words = [];
let v = 0;
while (v < 7) {
    let n = 0;
    while (n <= v) {
        words.push(vocab[v]);
        n = n + 1;
    }
    v = v + 1;
}
const line = words.join(" ");

const counts = new Map();
let round = 0;
while (round < 150000) {
    const parts = line.split(" ");
    let i = 0;
    while (i < parts.length) {
        const w = parts[i];
        counts.set(w, (counts.get(w) ?? 0) + 1);
        i = i + 1;
    }
    round = round + 1;
}

const items = Array.from(counts.entries());
items.sort((a, b) => b[1] - a[1]);
let j = 0;
while (j < items.length) {
    console.log(items[j][0], items[j][1]);
    j = j + 1;
}

File: bench/wordfreq/wordfreq.lua
-- Word frequency counting: string split, dict updates, and a sort with a key.
local vocab = {"alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta"}

local words = {}
local v = 1
while v <= 7 do
    local n = 0
    while n < v do
        words[#words + 1] = vocab[v]
        n = n + 1
    end
    v = v + 1
end
local line = table.concat(words, " ")

local function split(s, sep)
    local parts = {}
    local start = 1
    while true do
        local at = string.find(s, sep, start, true)
        if at == nil then
            parts[#parts + 1] = string.sub(s, start)
            return parts
        end
        parts[#parts + 1] = string.sub(s, start, at - 1)
        start = at + #sep
    end
end

local counts = {}
local round = 0
while round < 150000 do
    local parts = split(line, " ")
    local i = 1
    while i <= #parts do
        local w = parts[i]
        counts[w] = (counts[w] or 0) + 1
        i = i + 1
    end
    round = round + 1
end

local items = {}
for word, count in pairs(counts) do
    items[#items + 1] = {word, count}
end
table.sort(items, function(a, b) return a[2] > b[2] end)
local j = 1
while j <= #items do
    print(items[j][1] .. " " .. items[j][2])
    j = j + 1
end
